### PR TITLE
Issue/213 添加conv算子cpu/cuda实现

### PR DESCRIPTION
**Pull request by Graylatzhou**
image
image

### CODE REVIEW

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        return INFINI_STATUS_BAD_TENSOR_DTYPE;
+    }
+
+    auto result = ConvInfo::create(handle_, y_desc, x_desc, w_desc, b_desc,
+                                   pads, strides, dilations, n);
+
+    CHECK_RESULT(result);
+    size_t workspace_size = result->handler->workspace_size;
+    *desc_ptr = new Descriptor(
+        dtype, result.take(), workspace_size,
```
Comment (Ziminli) [inline src/infiniop/ops/conv/cuda/conv_cuda.cu:38]: (line 38) 用 `std::move`

File: src/infiniop/ops/conv/info.h
Comment (Ziminli) [inline src/infiniop/ops/conv/info.h:1]: 这里不太能这么设计，
1. `ConvInfo` 本身还好，但推荐统一用一个 meta mem pool 然后每个 meta（i.e., `input_dims`, `kernel_dims`, etc.) 用 offset 获取。可以参考 `ElementwiseInfo` 的设计；
2. 去除 `CudnnConvHandler`, 这个 cuda 特定的信息结构可以在 cuda 的 `Opaque` 里设计。`Opaque` 对外隐藏，各个后端平台可以在里面进行自己的设计，这里可以详细阅读 gemm.h 里的对该设计的解释。`ConvInfo` 里只应有通用的属性和信息，不应该包含例如 `CudnnConvHandler` 这种结构（即便用条件编译），这本身不是个好的设计。
3. 第二点中提及的设计形式目前 InfiniCore 还没有（没有使用 cudnn 的），但 conv 的具体实现（cudnn 调用和检查等）可以参考之前提及的 [operators 123 PR](https://github.com/InfiniTensor/operators/pull/123/files)。

File: src/infiniop/ops/conv/cpu/conv_cpu.cc [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+infiniStatus_t Descriptor::calculate(
+    void *workspace,
+    size_t workspace_size,
+    void *y,
+    const void *x,
+    const void *w,
+    const void *bias,
+    void *stream) const {
+    if (workspace_size < _workspace_size) {
+        return INFINI_STATUS_BAD_PARAM;
```
[the remark this answers is not in the document]
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cpu/conv_cpu.cc:314]: 已修正

File: src/infiniop/ops/conv/cpu/conv_cpu.cc [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    if (dtype != INFINI_DTYPE_F16 && dtype != INFINI_DTYPE_F32) {
+        return INFINI_STATUS_BAD_TENSOR_DTYPE;
+    }
+
+    auto result = ConvInfo::create(handle_, y_desc, x_desc, w_desc, b_desc,
+                                   pads, strides, dilations, n);
+    CHECK_RESULT(result);
+
+    size_t WorkSpaceSize = 0;
+    ConvInfo &info = *result;
```
[the remark this answers is not in the document]
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cpu/conv_cpu.cc:67]: 已修正

File: src/infiniop/ops/conv/info.h
Comment (Graylatzhou) [reply src/infiniop/ops/conv/info.h:1]: cudnn handle已经修改为通过opaque来管理

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        infiniDtype_t data_type,
+        cudnnDataType_t compute_type) {
+
+        bool is_1d_conv = (info.ndim() == 1);
+        int actual_tensor_ndim = is_1d_conv ? 4 : static_cast<int>(info.ndim() + 2);
+        int spatial_ndim_for_conv_desc = static_cast<int>(info.ndim());
+
+        if (is_1d_conv) {
+            spatial_ndim_for_conv_desc = 2;
+        }
```
Comment (Ziminli) [inline src/infiniop/ops/conv/cuda/conv_cuda.cu:67]: 像上面 `actual_tensor_ndim` 一样合成一行

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+        cudnnDataType_t cudnn_data_type;
+        if (data_type == INFINI_DTYPE_F16) {
+            cudnn_data_type = CUDNN_DATA_HALF;
+        } else if (data_type == INFINI_DTYPE_F32) {
+            cudnn_data_type = CUDNN_DATA_FLOAT;
+        } else {
+            return INFINI_STATUS_BAD_TENSOR_DTYPE;
+        }
```
Comment (Ziminli) [inline src/infiniop/ops/conv/cuda/conv_cuda.cu:126]: 用 `device::cuda::getCudnnDtype`

File: src/infiniop/ops/conv/info.h
Comment (Ziminli) [inline src/infiniop/ops/conv/info.h:1]: 从外部传入的 strides 要按 `ptrdiff_t` 来存取，只是到实际的某个平台实现里如果有需要（比如 `cudnn` 要 `int32_t` 之类的）再在那里转成对应的类型

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        if (conv_desc) {
+            cudnnDestroyConvolutionDescriptor(conv_desc);
+            conv_desc = nullptr;
+        }
+    }
+
+    infiniStatus_t GenHandler(
+        ConvInfo &info,
+        infiniDtype_t data_type,
+        cudnnDataType_t compute_type) {
```
Comment (Ziminli) [inline src/infiniop/ops/conv/cuda/conv_cuda.cu:59]: 这个函数的 body 部分太长了，适当进行一些抽象（比如下面一开始的所有元信息数组初始化可以抽象成一个，有无 bias 的 cudnn 调用各抽象成一个之类的，这个需要整理设计一下，需要尽量减少冗余），辅助函数设置为 `Opaque` 中的 private

**Review by Ziminli — CHANGES_REQUESTED** [review]

File: include/infiniop/ops/softmax.h
[the remark this answers is not in the document]
Comment (Graylatzhou) [reply include/infiniop/ops/softmax.h:1]: 已删除

File: src/infiniop/ops/conv/cuda/conv_cuda.cu (hunk 4 above)
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cuda/conv_cuda.cu:67]: 已修改

File: src/infiniop/ops/conv/cuda/conv_cuda.cu (hunk 5 above)
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cuda/conv_cuda.cu:126]: 已修改

File: src/infiniop/ops/conv/info.h
Comment (Graylatzhou) [reply src/infiniop/ops/conv/info.h:1]: 已修改

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+            b_desc = nullptr;
+        }
+        if (act_desc) {
+            cudnnDestroyActivationDescriptor(act_desc);
+            act_desc = nullptr;
+        }
+        if (conv_desc) {
+            cudnnDestroyConvolutionDescriptor(conv_desc);
+            conv_desc = nullptr;
+        }
```
[the remark this answers is not in the document]
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cuda/conv_cuda.cu:53]: 已实现

**Review by Ziminli — CHANGES_REQUESTED** [review]

File: src/infiniop/ops/conv/cuda/conv_cuda.cu [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+        other.act_desc = nullptr;
+        other.conv_desc = nullptr;
+        other.workspace_size = 0;
+    }
+
+    ~Opaque() {
+        CLEANUP_CUDNN_DESCRIPTORS();
+    }
+
+    infiniStatus_t GenHandler(ConvInfo &info,
```
[the remark this answers is not in the document]
Comment (Graylatzhou) [reply src/infiniop/ops/conv/cuda/conv_cuda.cu]: 已经修改，修改为了initializeCudnnContext
修改getStridesInfo() 和 stride_info() 返回ptrdiff_t

File: test/infiniop/conv.py [hunk 9]
```diff
@@ -95,31 +169,37 @@ def test(
...
+    x = torch.rand(x_shape, dtype=tensor_dtype).to(torch_device) * 0.001
+    w = torch.rand(w_shape, dtype=tensor_dtype).to(torch_device) * 0.001
     y = torch.zeros(
         inferShape(x.shape, w.shape, pads, strides, dilations), dtype=tensor_dtype
     ).to(torch_device)
+    bias = (
+        torch.rand(w.shape[0], dtype=tensor_dtype).to(torch_device) * 0.001
+        if w.shape[0] > 1
+        else None
+    )
```
Comment (Ziminli) [inline test/infiniop/conv.py:187]: 这里数据范围调成 0-0.001 是有什么考虑吗

**Review by Ziminli — CHANGES_REQUESTED** [review]

File: test/infiniop/conv.py (hunk 9 above)
Comment (PanZezhong1725) [reply test/infiniop/conv.py:187]: 如果怕溢出应该x和w有一个调小就可以了，两个都调太小了

Comment (Graylatzhou) [reply test/infiniop/conv.py:187]: 主要是怕溢出，如果是*0.1就会导致溢出

**Review by Ziminli — CHANGES_REQUESTED** [review]

**Comment by PanZezhong1725** [comment]
测试框架有重构，请rebase最新代码做相应更改

**Comment by Ziminli** [comment]
现在支持 bf16 了，麻烦 cpu 和 cuda 都适配一下 bf16，要改的应该不会很多
